Mathlib/Computability/ContextFreeGrammar.lean

```diff
@@ -0,0 +1,132 @@
+  /-- Nonterminal symbols (must not be present at the end of word being generated) -/
+  | nonterminal (n : N) : Symbol T N
+  -- If we have `[BEq T]` and `[BEq N]` then `[BEq (Symbol T N)]` is added.
+  deriving BEq
+
+/-- Rule that rewrites a single nonterminal to any list of symbols. -/
+structure ContextFreeRule (T : Type _) (N : Type _) where
+  /-- Input nonterminal a.k.a. left-hand side -/
+  input : N
+  /-- Output string a.k.a. right-hand side -/
+  output : List (Symbol T N)
+
+/-- Context-free grammar that generates words over the alphabet `T` (a type of terminals). -/
+structure ContextFreeGrammar (T : Type _) where
+  /-- Type of nonterminals -/
+  NT : Type _
+  /-- Initial nonterminal -/
+  initial : NT
+  /-- Rewrite rules -/
+  rules : List (ContextFreeRule T NT)
+
+variable {T : Type _}
+
+/-- One application of single context-free rule. -/
+inductive ContextFreeRule.RewritesTo {N : Type _} (r : ContextFreeRule T N) :
+    List (Symbol T N) → List (Symbol T N) → Prop
+  /-- The replacement is at the start of the remaining string. -/
+  | head (s : List (Symbol T N)) :
+      r.RewritesTo (Symbol.nonterminal r.input :: s) (r.output ++ s)
+  /-- There is a replacement later in the string. -/
+  | cons (x : Symbol T N) {s₁ s₂ : List (Symbol T N)} (hrs : RewritesTo r s₁ s₂) :
+      r.RewritesTo (x :: s₁) (x :: s₂)
+
+lemma ContextFreeRule.RewritesTo.exists_parts {N : Type _} {r : ContextFreeRule T N}
+    {u v : List (Symbol T N)} (hyp : r.RewritesTo u v) :
+    ∃ p : List (Symbol T N), ∃ q : List (Symbol T N),
+      u = p ++ [Symbol.nonterminal r.input] ++ q ∧ v = p ++ r.output ++ q := by
+  induction hyp with
+  | head xs =>
+    use [], xs
+    simp
+  | cons x _ ih =>
+    rcases ih with ⟨p', q', bef, aft⟩
+    use x :: p', q'
+    simp [bef, aft]
+
+lemma ContextFreeRule.rewritesTo_of_exists_parts {N : Type _}
+    (r : ContextFreeRule T N) (p q : List (Symbol T N)) :
+    r.RewritesTo (p ++ [Symbol.nonterminal r.input] ++ q) (p ++ r.output ++ q) := by
+  induction p with
+  | nil         => exact ContextFreeRule.RewritesTo.head q
+  | cons d l ih => exact ContextFreeRule.RewritesTo.cons d ih
+
+theorem ContextFreeRule.rewritesTo_iff {N : Type _} {r : ContextFreeRule T N}
+    (u v : List (Symbol T N)) :
+    r.RewritesTo u v ↔
+      ∃ p : List (Symbol T N), ∃ q : List (Symbol T N),
+        u = p ++ [Symbol.nonterminal r.input] ++ q ∧ v = p ++ r.output ++ q := by
+  constructor
+  · apply ContextFreeRule.RewritesTo.toParts
+  · rintro ⟨p, q, bef, aft⟩
+    rw [bef, aft]
+    apply ContextFreeRule.rewritesTo_of_parts
+
+namespace ContextFreeGrammar
+
+/-- One step of context-free transformation. -/
+def Produces (g : ContextFreeGrammar T) (u v : List (Symbol T g.NT)) : Prop :=
+  ∃ r ∈ g.rules, r.RewritesTo u v
+
+/-- Any number of steps of context-free transformation. -/
+def Derives (g : ContextFreeGrammar T) : List (Symbol T g.NT) → List (Symbol T g.NT) → Prop :=
+  Relation.ReflTransGen g.Produces
```

Comment:
```suggestion
abbrev Derives (g : ContextFreeGrammar T) : List (Symbol T g.NT) → List (Symbol T g.NT) → Prop :=
  Relation.ReflTransGen g.Produces
```
since the fact it's a closure may as well not be an implementation detail.